source/server/common.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/http/header_map.h"
+
+#include "external/envoy/source/common/singleton/const_singleton.h"
+
+#include "api/server/response_options.pb.h"
+
+namespace Nighthawk {
+namespace Server {
+
+namespace TestServer {
+
+class HeaderNameValues {
+public:
+  const Envoy::Http::LowerCaseString TestServerConfig{"x-nighthawk-test-server-config"};
+};
+
+using HeaderNames = Envoy::ConstSingleton<HeaderNameValues>;
+
+} // namespace TestServer
+
+class Utility {
+public:
+  /**
+   * Merges a json string containing configuration into a ResponseOptions instance.
+   *
+   * @param json Json-formatted seralization of ResponseOptions to merge into the configuration.
+   * @param config The target that the json string should be merged into.
+   * @param error_message Will contain an error message iff an error occurred.
```

Comment:
The problem described there is:

```
The first option probably doesn’t do what you want. 
If someone passes a Foo into MyFunc, the Foo will be copied by value into the absl::optional<Foo>, 
which will then be passed by reference into the function. 
If your goal was to avoid copying the Foo, you haven’t.
```

My intent here was that that we actually do copy by value, and that the optional takes ownership of the string we return.
This in turn makes it easy to manage the lifetime of the (dynamic) error string synthesized in the function itself.
If was hoping that by using the optional, it would also be intuitive that a caller may or may not get an error message.
I'm hoping that the error flow, where this copying will occur, will be a low-frequency event, and as such not hurt performance in the hot path.

With that as some more context, does that change how you feel about it?

Reply:
My understanding is that nighthawk follows envoy style, and envoy largely follows the google-published C++ style guide.
See https://github.com/envoyproxy/nighthawk/pull/390/files#r450301523.

We try to save absl::optional for use in optional input parameters, and pointers are used for output parameters.

(In this case, for me reading the code, I was confused why the code was accepting an optional error message and then immediately overriding it.  I had to jump to this comment to understand it was an output error message.)